### PR TITLE
Support any-dimensional memrefs in `AllocsToSLM` pass

### PR DESCRIPTION
**Pull request by dchigarev**
This PR adds support for memrefs of any dimensions as well as for thread indices of all 3 dimensions (thread_idx_z was not supported before)

This PR fixes the failure of the ROPE module caused by the `AllocsToSLM` pass

### CODE REVIEW

File: lib/gc/Transforms/GPU/AllocsToSLM.cpp [hunk 1]
```diff
@@ -80,22 +81,17 @@ struct ConvertAlloc : public OpRewritePattern<memref::AllocOp> {
...
-    int64_t newX = originalShape[0] * xI;
-    int64_t newY = originalShape[1] * yI;
-
-    SmallVector<int64_t> newShape = {newX, newY};
+    // Scale the allocation size (X dimension) by the number of threads in the
+    // work-group
+    int64_t newX =
+        originalShape[0] * blockSizes[0] * blockSizes[1] * blockSizes[2];
+    SmallVector<int64_t> newShape({newX});
+    newShape.append(originalShape.begin() + 1, originalShape.end());
```
Comment (dchigarev) [inline lib/gc/Transforms/GPU/AllocsToSLM.cpp:94]: The logic of scaling allocation size by the total number of threads was reworked. We now scale only the zero dimension of the memref:
```mlir
// num threads in the work group: X=2, Y=3, Z=4
// alloc before the pass
%slm_buff = memref.alloc() : memref<8x16xf16>

// alloc after the pass (only scaled the zero dimension)
%slm_buff_root = memref.alloc() : memref<192x16xf16, 3>
%x_offset = slm_chunk_x_size * (X_thread_idx * Y_block_size * Z_block_size + Y_thread_idx * Z_block_size + Z_thread_idx) = 8 * (X_thread_idx * 12 + Y_thread_idx * 4 + Z_thread_idx)
%slm_buff = memref.subview [%x_offset, 0] : memref<192x16xf16, 3> -> memref<8x16xf16, 3>
```

**Review by kurapov-peter — APPROVED** [review]
LG

File: lib/gc/Transforms/GPU/AllocsToSLM.cpp [hunk 2]
```diff
@@ -80,22 +81,17 @@ struct ConvertAlloc : public OpRewritePattern<memref::AllocOp> {
...
-    int64_t xI = xSz.value();
-    int64_t yI = ySz.value();
-    int64_t zI = zSz.value();
-
-    if (zI != 1)
-      return rewriter.notifyMatchFailure(
-          allocOp, "Only support 2D shared memory for now");
-
+    SmallVector<int64_t, 3> blockSizes = {xSz.value(), ySz.value(),
+                                          zSz.value()};
```
Comment (AndreyPavlenko) [inline lib/gc/Transforms/GPU/AllocsToSLM.cpp:85]: Do you really need a Vector herre? It's only used to get elements by intex and could be replaced with either local variables or int64_t[] array.

**Review by AndreyPavlenko — APPROVED** [review]

File: lib/gc/Transforms/GPU/AllocsToSLM.cpp (hunk 2 above)
Comment (dchigarev) [reply lib/gc/Transforms/GPU/AllocsToSLM.cpp]: changed vector to a static array here

**Pull request merged**